app/lib/domains.server.ts

```diff
@@ -0,0 +1,146 @@
+  name: string;
+  value: string;
+}
+
+interface Record {
+  id?: number;
+  username?: string;
+  description?: string;
+  course?: string;
+  ports?: string;
+}
+
+export const createUserDomain = async (data: Record & DnsRecord) => {
+  const existingRecords = await prisma.record.findMany({
+    where: {
+      name: data.name,
+      type: data.type,
+      value: data.value,
+    },
+  });
+
+  if (existingRecords.length > 0) {
+    throw new Error('Record already exists');
+  }
+
+  const changeId = await dns.createRecord(data.type, data.name, data.value);
+
+  if (!changeId) {
+    throw new Error('Could not get ChangeID from creating a record in Route 53');
+  }
+
+  if (!data.username) {
+    throw new Error('Username is not provided');
+  }
+
+  const result = await prisma.record.create({
+    data: {
+      username: data.username,
+      type: data.type,
+      name: data.name,
+      value: data.value,
+      expiresAt: changeMonthFromNow(6),
+      status: RecordStatus.pending,
+    },
+  });
+
+  if (!result) {
+    throw new Error('Could not create a record in DB');
+  }
+  return result;
+};
+
+export const updateUserDomain = async (data: Record & DnsRecord) => {
+  const changeId = await dns.upsertRecord(data.type, data.name, data.value);
+
+  if (!changeId) {
+    throw new Error('Could not get ChangeID from updating a data in Route 53');
+  }
+
+  const result = await prisma.record.update({
+    where: {
+      id: data.id,
+    },
+    data: {
+      name: data.name,
+      type: data.type,
+      value: data.value,
+      description: data.description,
+      course: data.course,
+      ports: data.ports,
+      expiresAt: changeMonthFromNow(6),
+      status: RecordStatus.pending,
+    },
+  });
+
+  if (!result) {
+    throw new Error('Could not update the record in DB');
+  }
+  return result;
+};
+
+export const deleteUserDomain = async (data: Record & DnsRecord) => {
+  const changeId = await dns.deleteRecord(data.type, data.name, data.value);
+
+  if (!changeId) {
+    throw new Error('Could not get ChangeID from deleting a record in Route 53');
+  }
+
+  const result = await prisma.record.delete({
+    where: {
+      id: data.id,
+    },
+  });
+
+  if (!result) {
+    throw new Error('Could not delete the record in DB');
+  }
+  return result;
+};
+
+export const removeIfExpired = async (data: Record & DnsRecord) => {
+  const recordFromId = await prisma.record.findUnique({
+    where: {
+      id: data.id,
+    },
+  });
+
+  if (!recordFromId) {
+    throw new Error('Record with provided ID does not exist');
+  }
+
+  if (isDomainExpired(recordFromId.expiresAt)) {
+    await dns.deleteRecord(data.type, data.name, data.value);
+
+    const result = await prisma.record.delete({
+      where: {
+        id: data.id,
+      },
+    });
+
+    if (!result) {
+      throw new Error('Could not delete the record in DB');
+    }
+    return result;
+  }
+};
+
+export const isDomainExpired = (expiresAt: Date) => {
```

Comment:
This name is confusing, since you aren't passing a domain, but rather a date.  `isExpired`?